L10-c++-reactive-streams.md

```diff
@@ -0,0 +1,387 @@
+
+It is being used extensively by [Netflix](https://medium.com/netflix-techblog/reactive-programming-in-the-netflix-api-with-rxjava-7811c3a1496a), [Akka](http://doc.akka.io/docs/akka/2.5.4/scala/stream/stream-introduction.html), Github and many other companies. The upcoming version 5 of the widely used Spring framework [is based on Reactive Streams](https://spring.io/blog/2016/07/28/reactive-programming-with-spring-5-0-m1). A precursor to Reactive Streams is in [LINQ in C#](https://msdn.microsoft.com/en-us/library/hh242977(v=vs.103).aspx). Reactive Streams [is being standardized in Java 9](https://community.oracle.com/docs/DOC-1006738). With [ReactiveX](http://reactivex.io/languages.html) this programming model is available in lots of programming languages (albeit not always with backpressure support).
+
+There is a lot of material on the Internet that covers Reactive Streams from many different angles. [ReactiveX](http://reactivex.io/) has lots of good documentation and [Marble diagrams](http://rxmarbles.com/) are a wonderful way to describe what Rx operators do.
+
+
+## Implementation
+
+A prototype of what is suggested in this RFC can be found [here](https://github.com/per-gron/shuriken/tree/master/src/rs-grpc). It implements the most important features, but it is still quite incomplete.
+
+Implementation of this RFC would involve:
+
+* Prepare the [rs](https://github.com/per-gron/shuriken/tree/master/src/rs) library so that it can be accepted as gRPC code.
+* Improve the prototype library so that it can be used for experimental purposes (fix memory leaks etc).
+* Write a protobuf code generator to be able to remove the most expensive C++ templates from the prototype and to make the API nicer.
+* Implement missing features in the prototype: Finish cancellation support, add support for timeouts, metadata etc.
+* Polish API and test.
+* Evangelize.
+
+The author could do much of this work, but it would be awesome to get help.
+
+
+## Open issues
+
+* The RPC handler methods might need to take one or more additional parameters to handle gRPC features like metadata.
+* This proposal glosses over many details; the API needs to be fleshed out more.
+* There are certain details in the current gRPC C++ API that seem to be missing: For example, the author could not find how get a signal in the server when a unary RPCs is cancelled.
+* The rs library is a fairly minimal implementation of Reactive Streams. One thing that many Rx libraries have that rs does not is a concept for a Publisher that emits only one value. This would increase the complexity of the library a bit, but it could be worth it for the additional type safety and performance benefits it would bring.
+* [rs](https://github.com/per-gron/shuriken/tree/master/src/rs) breaks Google's code style in the following ways. I am unsure about how exactly to fix this.
+  * **It uses C++14:** This is mostly for ease of implementation; it should be possible to backport to C++11.
```

Comment:
We'll need to stay with C++11 for some time.

Reply:
I understand. If people actually are interested in this I will definitely consider backporting it. It's just a bit of a pain to do and parts of the implementation would be uglier/harder to read.

Reply:
I have added a bullet point under the "Implementation" section to backport to C++11.

With this I consider this issue resolved, let me know if you disagree.